Zip: Add support for modification time and date

Ollrogge opened the pull request:
This PR adds support for the modification date and time in zip archives. This means that after extracting the archive the file modification times are the same as the ones of the original files.

timschumi commented on Userland/Libraries/LibArchive/Zip.cpp:
```diff
@@ -90,7 +91,7 @@ ErrorOr<bool> Zip::for_each_member(Function<IterationDecision(ZipMember const&)>
         member.uncompressed_size = central_directory_record.uncompressed_size;
         member.crc32 = central_directory_record.crc32;
-        member.is_directory = central_directory_record.external_attributes & zip_directory_external_attribute || member.name.bytes_as_string_view().ends_with('/'); // FIXME: better directory detection
+        member.is_directory = central_directory_record.external_attributes & zip_directory_external_attribute || Core::File::is_directory(member.name.to_deprecated_string());
```
A directory existing on-disk that (by chance) has the same name as the current member doesn't automatically make the current member a directory. This presumably also makes parsing a lot slower, since we are accessing the disk for each member.

Ollrogge replied:
Okay, will remove this change

timschumi commented on Userland/Utilities/unzip.cpp:
```diff
@@ -17,6 +17,33 @@
 #include <LibCrypto/Checksum/CRC32.h>
 #include <sys/stat.h>
 
+static time_t dos_to_time_t(u16 dos_time, u16 dos_date)
+{
+    struct tm tm;
+
+    tm.tm_isdst = -1;
+    tm.tm_year = ((dos_date >> 9) & 127) + 1980 - 1900;
+    tm.tm_mon = ((dos_date >> 5) & 15) - 1;
+    tm.tm_mday = dos_date & 31;
+
+    tm.tm_hour = (dos_time >> 11) & 31;
+    tm.tm_min = (dos_time >> 5) & 63;
+    tm.tm_sec = (dos_time << 1) & 62;
+
+    return mktime(&tm);
+}
```
This should be somewhere inside `LibArchive`. It also reminds me a lot of the time translation in our FAT filesystem implementation, maybe there is some opportunity to share code.

Ollrogge replied:
It actually works exactly the same like the `FATPacked*`  times. I oversaw this unfortunately. Therefore the code can be shared very well.
Do you have any recommendation of how to do this in the best way ? Should I create a class called e.g. `MSPackedDate` (`MS` = Microsoft) that handles the translations?

timschumi replied:
The more interesting question is _where_ to put it. I thought of just adding conversion functions to `AK::Time`, but that is apparently for durations, not for an absolute date-time. Maybe we should just create `AK::DateTime` to just have _somewhere_ to put it and look what others think about it.

timschumi replied:
Eh, looks like `AK::Time` _does_ support absolute datetimes, so we would just add conversion methods to/from the packed DOS format there.

Ollrogge replied:
Problem with `AK::Time` is that is has no support to translate an absolute datetime back to a date

Ollrogge replied:
Decided to make a dedicated file and host the conversion functions there since they didn't seem to fit nicely into `AK/Time` since it doesn't support conversion from absolute times to date.

If this approach is alright, I would continue to refactor the FAT code to use the conversion functions as well.

timschumi commented on Userland/Utilities/zip.cpp:
```diff
@@ -8,12 +8,21 @@
 #include <LibCore/System.h>
 #include <LibCrypto/Checksum/CRC32.h>
 
+static void time_t_to_dos(time_t sec, u16* dos_time, u16* dos_date)
+{
+    auto date = Core::DateTime::from_timestamp(sec);
+
+    *dos_time = date.hour() << 11 | date.minute() << 5 | date.second() >> 1;
+    *dos_date = (date.year() - 1980) << 9 | date.month() << 5 | date.day();
+}
+
```
Same here.

timschumi commented on Userland/Utilities/unzip.cpp:
```diff
@@ -17,6 +17,33 @@
+    return mktime(&tm);
+}
+
+static int adjust_modification_time(Archive::ZipMember const& zip_member)
+{
+    time_t m_time = dos_to_time_t(zip_member.modification_time, zip_member.modification_date);
+    struct utimbuf buf {
+        .actime = m_time,
+        .modtime = m_time
+    };
+
+    return utime(zip_member.name.to_deprecated_string().characters(), &buf);
+}
```
`Core::System::utime()` and proper error propagation please.

timschumi commented on Userland/Utilities/unzip.cpp:
```diff
@@ -147,5 +179,20 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
+        return 1;
+    }
+
+    success = TRY(zip_file->for_each_member([&](auto zip_member) {
+        if (zip_member.is_directory) {
+            if (adjust_modification_time(zip_member) < 0) {
+                warnln("Failed setting modification time for directory {}", zip_member.name);
+                return IterationDecision::Break;
+            }
+        }
+
+        return IterationDecision::Continue;
+    }));
```
Why are we iterating the whole ZIP a second time?

Ollrogge replied:
Regarding this and the next two comments: If I set the modification date of a directory in `unpack_zip_member` it will get overwritten when files are created in this directory. Therefore, the original modification time for a directory needs to be set after all files have been unpacked. 
This was the best solution I came up with to do this.

timschumi replied:
We should probably store a list of directories (with the necessary attributes included) while parsing the file the first time, and use that to restore the settings instead.

Ollrogge replied:
Decided to make a dedicated file and host the conversion functions there since they didn't seem to fit nicely into `AK/Time` since it doesn't support conversion from absolute times to date.

If this approach is alright, I would continue to refactor the FAT code to use the conversion functions as well.

Ollrogge replied:
Done

timschumi commented on Userland/Utilities/unzip.cpp:
```diff
@@ -147,5 +179,20 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
+    }
+
+    success = TRY(zip_file->for_each_member([&](auto zip_member) {
+        if (zip_member.is_directory) {
```
Why are we doing this for directories only?

timschumi commented on Userland/Utilities/unzip.cpp:
```diff
@@ -147,5 +179,20 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
+
+    success = TRY(zip_file->for_each_member([&](auto zip_member) {
+        if (zip_member.is_directory) {
+            if (adjust_modification_time(zip_member) < 0) {
```
Why don't we just put this into `unpack_zip_member`?

timschumi commented on Userland/Utilities/zip.cpp:
```diff
@@ -58,6 +67,10 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
         Archive::ZipMember member {};
         member.name = TRY(String::from_deprecated_string(canonicalized_path));
 
+        struct stat st;
+        fstat(file->fd(), &st);
```
`Core::System::fstat()`

timschumi commented on Userland/Utilities/zip.cpp:
```diff
@@ -85,6 +98,11 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
         member.crc32 = 0;
         member.is_directory = true;
+
+        struct stat st;
+        stat(canonicalized_path.characters(), &st);
```
`Core::System::stat()`

drunderscore reviewed (changes requested):
Same issues as Tim -- the first commit is certainly not correct.
Additionally, `Zip` is not a valid category to use for the commit title -- instead use `LibArchive`

Ollrogge commented:
@timschumi  Can you take another look ?

ADKaster commented on AK/DOSPackedTime.cpp:
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (c) 2020, the SerenityOS developers.
```
Was this file created 3 years ago?

timschumi replied:
Since we are looking to move the definitions for the FAT date and time format into `AK` to share them, and this is very clearly based on the existing structs from `Kernel/FileSystem/FATFS/Definitions.h`, we should probably 'inherit' from that copyright (which is 2022 by Undefine). Feel free to add on yours too, of course, but with the correct year.

Also, please make the FAT implementation in the Kernel use these structs and methods here (if possible), and remove the now-duplicated ones in `Kernel/FileSystem/FATFS/Definitions.h` and `Kernel/FileSystem/FATFS/Inode.cpp`.

ADKaster commented on AK/DOSPackedTime.h:
```diff
@@ -0,0 +1,38 @@
+DOSPackedDate to_packed_dos_date(unsigned, unsigned, unsigned);
+DOSPackedTime to_packed_dos_time(unsigned, unsigned, unsigned);
+
+}
```
New AK types should be conditionally `using`'d into the global namespace wrapped by `#if USING_AK_GLOBALLY` to follow the pattern of the rest of the library.

ADKaster commented on AK/DOSPackedTime.cpp:
```diff
@@ -0,0 +1,35 @@
+DOSPackedTime to_packed_dos_time(unsigned hour, unsigned minute, unsigned second)
+{
+    DOSPackedTime time;
+    time.value = hour << 11 | minute << 5 | second >> 1;
```
Can we write to the named bitfields (year, month, date, hour minute, second) rather than writing to the value field and shifting the bits into place? It would be a lot easier to read and I trust the compiler to coalesce the writes.

ADKaster commented on Userland/Libraries/LibArchive/Zip.h:
```diff
@@ -112,8 +113,8 @@ struct [[gnu::packed]] CentralDirectoryRecord {
-    u16 modification_time;
-    u16 modification_date;
+    AK::DOSPackedTime modification_time;
+    AK::DOSPackedDate modification_date;
```
These should use `DOSPackedTime` without the `AK::` namespace prefix after the names are exported into the global namespace.

ADKaster commented on Userland/Utilities/unzip.cpp:
```diff
@@ -17,6 +18,18 @@
+static ErrorOr<void> adjust_modification_time(Archive::ZipMember const& zip_member)
+{
+    auto time = AK::time_from_packed_dos(zip_member.modification_date, zip_member.modification_time);
+    auto ts = time.to_timespec();
```
Did we need to create a timespec here if we're only caring about seconds? we can probably just slice off the seconds with `to_seconds`

```
auto seconds = static_cast<time_t>(time.to_seconds());
```

Perhaps a future refactor could name `static_cast<time_t>(to_seconds())` as `static time_t Time::to_posix_time()`

ADKaster reviewed (commented):
These changes are looking pretty slick! A few comments about style when adding AK types

Ollrogge commented:
@ADKaster @timschumi Addressed the latest comments. However I couldn't find a way how to easily test the FATFS code to see if I may have broken  something.

timschumi commented on AK/DOSPackedTime.cpp:
```diff
@@ -13,21 +13,25 @@ Time time_from_packed_dos(DOSPackedDate date, DOSPackedTime time)
         return Time();
 
-    return Time::from_timestamp(1980 + date.year, date.month, date.day, time.hour, time.minute, time.second * 2, 0);
+    return Time::from_timestamp(1980 + date.year, date.month, date.day, time.hour, time.minute, time.second << 1, 0);
```
`<< 1` should continue to be `* 2` here.

timschumi commented on AK/DOSPackedTime.cpp:
```diff
@@ -13,21 +13,25 @@ Time time_from_packed_dos(DOSPackedDate date, DOSPackedTime time)
-    time.value = hour << 11 | minute << 5 | second >> 1;
+    time.hour = hour;
+    time.minute = minute;
+    time.second = second >> 1;
```
`>> 1` should be `/ 2` instead.

timschumi commented on Kernel/FileSystem/FATFS/Inode.cpp:
```diff
@@ -186,14 +186,6 @@ ErrorOr<NonnullOwnPtr<KString>> FATInode::compute_filename(FATEntry& entry, Vect
-    if (date.value == 0)
-        return Time();
-
-    return Time::from_timestamp(first_fat_year + date.year, date.month, date.day, time.hour, time.minute, time.second * 2, 0);
```
`first_fat_year` seems to be unused now. However, I'd onsider to carry that over to the AK implementation, replacing the `1980` magic number.

timschumi reviewed (changes requested):
Please squash in `fixup` commits instead of leaving them separate. I'd also squash together the Kernel and AK commits, since we essentially move the packed date/time implementation from the Kernel to AK.

Ollrogge commented:
> Please squash in `fixup` commits instead of leaving them separate. I'd also squash together the Kernel and AK commits, since we essentially move the packed date/time implementation from the Kernel to AK.

Regarding the `fixup` commits: Should they always be squashes directly ? I always thought it is easier for reviewers to check the addressed comments in standalone fixup commits.

timschumi commented:
> Should they always be squashes directly ? I always thought it is easier for reviewers to check the addressed comments in standalone fixup commits.

Force-pushes are pretty readable, as long as you don't combine any actual changes with rebasing onto latest `master`. When in doubt, we (or at least I) just check the chain of commits again instead of relying on the "Compare" button that GitHub provides when someone force-pushes.

timschumi commented on Userland/Utilities/unzip.cpp:
```diff
@@ -140,12 +160,26 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
+        return 1;
+    }
+
+    for (auto& directory : zip_directories) {
+        if (adjust_modification_time(directory).is_error()) {
+            warnln("Failed setting modification time for directory {}", directory.name);
+            return 1;
+        }
+    }
```
Don't we have to iterate the directories in the reverse order here? I'd assume that the top-most directory would be listed first in the file, so we'd have to set its modification time last.

Ollrogge replied:
I don't know if I get your point. It shouldn't matter in which order we go through the saved `ZipMember` entries and set the modication times for a file based on the path stored in `zip_member.name`.

timschumi replied:
Won't setting the modification time of a child directory also update the modification date of the parent directory (as in, "update to the current time", not "update to whatever value was given")? If that isn't the case then don't mind me.

Ollrogge replied:
No this should only happen when subdirs / files are added / removed.

Made one more small change to only add a directory to the vector if unpacking was actually successful.
Should all be good now :)

timschumi reviewed (commented):
Looking pretty good, I just have one last question.

Ollrogge commented:
Oversaw that I had conflicts. All lights green now :)

AtkinsSJ commented on AK/DOSPackedTime.h:
```diff
@@ -0,0 +1,48 @@
+inline constexpr u16 first_dos_year = 1980;
+
+Time time_from_packed_dos(DOSPackedDate, DOSPackedTime);
+DOSPackedDate to_packed_dos_date(unsigned, unsigned, unsigned);
+DOSPackedTime to_packed_dos_time(unsigned, unsigned, unsigned);
```
Please name your parameters here, since it's not obvious from the types which is which.

AtkinsSJ commented on AK/CMakeLists.txt:
```diff
@@ -5,6 +5,7 @@ set(AK_SOURCES
     DeprecatedFlyString.cpp
     DeprecatedString.cpp
     Error.cpp
+    DOSPackedTime.cpp
```
Try to keep this in alphabetical order please. :^)

AtkinsSJ reviewed (changes requested):
Just some nit-picks.
